### PR TITLE
fix: tkps grad accu miscalculation

### LINKED ISSUE #3688: `tokens/train_per_sec_per_gpu` miscount with gradient accumulation?
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

`tokens/train_per_sec_per_gpu` should report trainable-token throughput per GPU in a way that accounts for:

- masked labels (`labels == -100`)
- DDP/world size
- `gradient_accumulation_steps`
- `logging_steps`

The cumulative counters appear correct:

- `tokens/trainable` counts `labels != -100`
- `tokens/total` counts total label tensor slots

The per-second throughput should be consistent with those counters.

### Current behaviour

`tokens/train_per_sec_per_gpu` appears to underreport throughput when `gradient_accumulation_steps > 1`.

Relevant code path:

- `core/trainers/base.py`: `trainable_tokens = (labels != -100).sum()`
- `core/trainers/base.py`: `state.tokens["trainable_tokens"] = trainable_tokens`
- `utils/callbacks/tokens_per_second.py`: computes throughput from `state.tokens["trainable_tokens"]`
- `core/trainers/base.py`: divides `last_tokens_per_second` by `logging_steps`

It looks like `trainable_tokens` is overwritten per microbatch rather than accumulated across the optimizer step or logging window.

Example from a current run:

```text
gradient_accumulation_steps: 8
world_size: 2
logging_steps: 10

0%| | 29/25748 [15:32<..., 17.16s/it]
{
  'loss': '0.09934',
  'tokens/train_per_sec_per_gpu': '11.32',
  'tokens/total': 1290025,
  'tokens/trainable': 301548,
  'epoch': '0.001165'
}
```

At that point:

```text
301,548 trainable tokens / 2 GPUs / 932 sec = ~162 trainable tokens/sec/GPU
```

The previous log line had:

```text
tokens/trainable: 217252
```

So the logging window processed:

```text
301,548 - 217,252 = 84,296 trainable tokens
84,296 / 2 GPUs / roughly 169-190 sec = ~222-249 trainable tokens/sec/GPU
```

The reported `11.32` does not match either cumulative throughput or logging-window throughput.

### Steps to reproduce

1. Run causal training with token throughput logging enabled. `include_tkps` defaults to `true`, so this may happen even if it is not explicitly present in the YAML.
2. Set `gradient_accumulation_steps > 1` and `logging_steps > 1`.
3. Compare logged `tokens/train_per_sec_per_gpu` against `tokens/trainable` deltas and elapsed wall time.
4. Observe that `tokens/trainable` and `tokens/total` look cumulative/correct, but `tokens/train_per_sec_per_gpu` is much lower than the implied throughput.

### Config yaml

```yaml
gradient_accumulation_steps: 8
micro_batch_size: 1
logging_steps: 10
```

### Possible solution

Option A: logging-window average

```text
trainable_per_sec_per_gpu = delta(tokens/trainable) / data_parallel_world_size / delta_seconds
total_per_sec_per_gpu = delta(tokens/total) / data_parallel_world_size / delta_seconds
```

This averages over the current logging window, e.g. 10 optimizer steps when `logging_steps: 10`.

Option B: latest optimizer-step throughput

Accumulate trainable/total tokens across all gradient-accumulation microbatches for the current optimizer step, then divide by that optimizer step's wall time and data-parallel world size.

Option C: log both

```text
tokens/trainable_per_sec_per_gpu
tokens/trainable_per_sec_per_gpu_step
tokens/total_per_sec_per_gpu
tokens/total_per_sec_per_gpu_step
```

If you'd like I can prepare a patch for A, B, C or otherwise. 

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main - most recent commit

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

- Fixes https://github.com/axolotl-ai-cloud/axolotl/issues/3688
- Refactors our callback and logging implementation given how it's scattered across two places.

### CODE REVIEW

File: src/axolotl/core/trainers/base.py [hunk 1]
```diff
@@ -755,15 +761,27 @@ def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
...
+                curr_trainable = float(self.state.tokens["trainable"].item())
+                elapsed = (
+                    now - self._tkps_prev_time
+                    if self._tkps_prev_time is not None
+                    else 0.0
+                )
+                rate = trainable_tokens_per_sec_per_gpu(
+                    self._tkps_prev_trainable,
+                    curr_trainable,
+                    get_world_size(),
```
Comment (winglian) [inline src/axolotl/core/trainers/base.py:775]: is world size correct since it depends on parallelisms too?

Comment (NanoCode012) [reply src/axolotl/core/trainers/base.py:775]: That was the intention. Prev, I think we divided by DP only, but this should account for EP/CP, so we can see actual throughput across num gpus per our naming `tokens per gpu`

**Review by winglian — APPROVED** [review]
